debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerChangeEventSourceCoordinator.java

```diff
@@ -0,0 +1,111 @@
+                                                 ChangeEventSourceMetricsFactory changeEventSourceMetricsFactory, EventDispatcher<?> eventDispatcher,
+                                                 DatabaseSchema<?> schema,
+                                                 Clock clock) {
+        super(previousOffsets, errorHandler, connectorType, connectorConfig, changeEventSourceFactory,
+                changeEventSourceMetricsFactory, eventDispatcher, schema);
+        this.clock = clock;
+        this.pollInterval = connectorConfig.getPollInterval();
+    }
+
+    @Override
+    protected void doSnapshot(CdcSourceTaskContext taskContext, SnapshotChangeEventSource<SqlServerPartition, SqlServerOffsetContext> snapshotSource,
+                              Offsets<SqlServerPartition, SqlServerOffsetContext> previousOffsets, AtomicReference<LoggingContext.PreviousContext> previousLogContext,
+                              ChangeEventSourceContext context)
+            throws InterruptedException {
+        Offsets<SqlServerPartition, SqlServerOffsetContext> streamingOffsets = Offsets.of(new HashMap<>());
+
+        for (Map.Entry<SqlServerPartition, SqlServerOffsetContext> entry : previousOffsets) {
+            SqlServerPartition partition = entry.getKey();
+            SqlServerOffsetContext previousOffset = entry.getValue();
+
+            SnapshotResult<SqlServerOffsetContext> snapshotResult = doSnapshot(snapshotSource, context, partition, previousOffset);
+
+            if (snapshotResult.isCompletedOrSkipped()) {
+                streamingOffsets.getOffsets().put(partition, snapshotResult.getOffset());
+            }
+        }
+
+        previousLogContext.set(taskContext.configureLoggingContext("streaming"));
+        streamEvents(context, streamingOffsets);
+    }
+
+    private void streamEvents(ChangeEventSourceContext context, Offsets<SqlServerPartition, SqlServerOffsetContext> streamingOffsets)
+            throws InterruptedException {
+
+        // TODO: Determine how to do incremental snapshots with multiple partitions but for now just use the first offset
+        SqlServerOffsetContext offsetContext = streamingOffsets.getOffsets().values().stream().findFirst().get();
+
+        initStreamEvents(offsetContext);
+        final Metronome metronome = Metronome.sleeper(pollInterval, clock);
+
+        LOGGER.info("Starting streaming");
+
+        while (running) {
+            boolean streamedEvents = false;
+            for (Map.Entry<SqlServerPartition, SqlServerOffsetContext> entry : streamingOffsets) {
+                SqlServerPartition partition = entry.getKey();
+                SqlServerOffsetContext previousOffset = entry.getValue();
+
+                if (running) {
+                    streamingSource.execute(context, partition, previousOffset);
+                    if (((SqlServerStreamingChangeEventSource) streamingSource).eventsStreamed(partition)) {
```

Comment:
It is probably better to discuss during today call but I must admit that I am not really happy about the current approach when this method is called and then another method provides the results. It feels a bit convoluted and hacky to me.

I'd prefer either to extend `StreamingChangeEventSource` with another method so the implementors will either provide fully managed handling or per-loop handling (proper name is needed) or there should be common `StreamingChangeEventSource` super interface and two specialized interfaces that would support either one or the other behaviour.

@gunnarmorling Your thoughts?

Reply:
@jpechane Thanks for the feedback. I went ahead and implemented the later suggestion.